### PR TITLE
[dv/kmac] functional coverage plan

### PR DESCRIPTION
**Pull request by udinator**
add the initial coverage plan for KMAC.

Signed-off-by: Udi Jonnalagadda <udij@google.com>

### CODE REVIEW

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 1]
```diff
@@ -158,4 +158,103 @@
...
+      desc: '''
+            Covers that the msgfifo has been written using all possible TLUL masks.
+            '''
+    }
+    {
+      name: msgfifo_status_cg
+      desc: '''
+            Covers that all possible fifo statuses have been seen in various operating states
+            (various fifo depths, full, empty).
+    }
```
Comment (sriyerg) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:204]: Missing `'''`.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 2]
```diff
@@ -158,4 +158,103 @@
...
+    {
+      name: msgfifo_write_mask_cg
+      desc: '''
+            Covers that the msgfifo has been written using all possible TLUL masks.
+            '''
+    }
+    {
+      name: msgfifo_status_cg
+      desc: '''
+            Covers that all possible fifo statuses have been seen in various operating states
```
Comment (sriyerg) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:202]: Perhaps enumerate the operating states here which will be crossed with the fifo levels. Also 'level' is more apt than 'status'?

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 3]
```diff
@@ -158,4 +158,103 @@
...
+            of scenarios:
+            - various msgfifo status (full/empty/in between)
+            - while keccak rounds are currently active/inactive
+            '''
+    }
+    {
+      name: sideload_cg
+      desc: '''
+            Covers that the KMAC sees scenarios where the sideloaded key is provided and should be
+            used (en_sideload==1, app_mode==AppKeymgr), and scenarios where the sideloaded key is
```
Comment (sriyerg) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:230]: Nit: use backticks for keywords

**Review by sriyerg — COMMENTED** [review]
Thanks Udi! Left a few comments, mainly with suggestions to make what is covered a bit more explicit / clear.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 4]
```diff
@@ -158,4 +158,103 @@
       tests: ["{name}_throughput"]
     }
   ]
+
+  covergroups: [
+    {
+      name: config_cg
+      desc: '''
+            Covers that all valid configuration settings for the KMAC have been tested.
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:166]: done.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 5]
```diff
@@ -158,4 +158,103 @@
...
+            '''
+    }
+    {
+      name: msg_len_cg
+      desc: '''
+            Covers various input message length ranges, to ensure that KMAC can operate successfully
+            on different sized messages.
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:173]: done.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 6]
```diff
@@ -158,4 +158,103 @@
...
+            '''
+    }
+    {
+      name: digest_len_cg
+      desc: '''
+            Covers  various output digest length ranges, to ensure that KMAC can successfully output
+            various digest lengths.
+            Note that this only applies to XOF functions, as SHA3 functions have a fixed output
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:181]: sure - I made the cg name a bit more explicit but still added a bit of description.

File: hw/ip/kmac/data/kmac_base_testplan.hjson (hunk 1 above)
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:204]: done.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 7]
```diff
@@ -158,4 +158,103 @@
...
+      name: msg_len_cg
+      desc: '''
+            Covers various input message length ranges, to ensure that KMAC can operate successfully
+            on different sized messages.
+            '''
+    }
+    {
+      name: digest_len_cg
+      desc: '''
+            Covers  various output digest length ranges, to ensure that KMAC can successfully output
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:179]: done (i think?)

File: hw/ip/kmac/data/kmac_base_testplan.hjson (hunk 2 above)
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:202]: done, thanks!

File: hw/ip/kmac/data/kmac_base_testplan.hjson (hunk 3 above)
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:230]: done, thanks!

**Review by matutem — APPROVED** [review]
LGTM

**Comment by udinator** [comment]
Have reviewed this coverage plan with @eunchan - the opinion is that this plan is fine as is, no additions need to be made.

**Review by eunchan — APPROVED** [review]

**Pull request merged**